OptimizelySDK/Odp/OdpManager.cs

```diff
@@ -0,0 +1,296 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using OptimizelySDK.ErrorHandler;
+using OptimizelySDK.Logger;
+using OptimizelySDK.Odp.Entity;
+using System;
+using System.Collections.Generic;
+
+namespace OptimizelySDK.Odp
+{
+    /// <summary>
+    /// Concrete implementation to orchestrate segment manager, event manager, and ODP config
+    /// </summary>
+    public class OdpManager : IOdpManager, IDisposable
+    {
+        /// <summary>
+        /// Denotes if ODP Manager is meant to be handling ODP communication
+        /// </summary>
+        private bool _enabled;
+
+        /// <summary>
+        /// Configuration used to communicate with ODP
+        /// </summary>
+        private volatile OdpConfig _odpConfig;
+
+        /// <summary>
+        /// Manager used to handle audience segment membership
+        /// </summary>
+        public IOdpSegmentManager SegmentManager { get; private set; }
+
+        /// <summary>
+        /// Manager used to send events to ODP 
+        /// </summary>
+        public IOdpEventManager EventManager { get; private set; }
+
+        /// <summary>
+        /// Logger used to record messages that occur within the ODP client
+        /// </summary>
+        private ILogger _logger;
+
+        /// <summary>
+        /// Update the settings being used for ODP configuration and reset/restart dependent processes
+        /// </summary>
+        /// <param name="apiKey">Public API key from ODP</param>
+        /// <param name="apiHost">Host portion of the URL of ODP</param>
+        /// <param name="segmentsToCheck">Audience segments to consider</param>
+        /// <returns>True if settings were update otherwise False</returns>
+        public bool UpdateSettings(string apiKey, string apiHost, List<string> segmentsToCheck)
+        {
+            var newConfig = new OdpConfig(apiKey, apiHost, segmentsToCheck);
+            if (_odpConfig.Equals(newConfig))
+            {
+                return false;
+            }
+
+            _odpConfig = newConfig;
+
+            // flush old events using old odp publicKey (if exists) before updating odp key.
+            // NOTE: It should be rare but possible that odp public key is changed for the same datafile (sdkKey).
+            // - Try to send all old events with the previous public key.
+            // - If it fails to flush all the old events here (network error), remaining events will be discarded.
+            EventManager.Flush();
```

Comment:
How can we guarantee all pending events use the old ODPConfig with the non-blocking flush?

Reply:
That's a good consideration. I think this `.Flush()` should grab a copy of the `_odpConfig` in the `OdpEventManager` and inject it into the `_flushSignal` object.  The secondary thread would then have the config to use for flush requests.

Does this seem like a suitable solution, or is this a corner case?  
How does Swift (or Python) handle this situation?
What's your direction?

Reply:
@mikechu-optimizely  We can consider -
1. (swift) the current key and host are captures in the flush thread (so they can be used until flushing completed) before updating odpConfig or
2. (python) delay config update until flush is completed with the current key and host.